File: AndroidVideoView/app/src/main/java/com/bowoon/android/android_videoview/video/VideoPlayerActivity.java
package com.bowoon.android.android_videoview.video;

import android.app.Activity;
import android.app.AlertDialog;
import android.content.Intent;
import android.content.pm.ActivityInfo;
import android.content.res.Configuration;
import android.media.MediaPlayer;
import android.os.Bundle;
import android.os.Handler;
import android.support.annotation.Nullable;
import android.util.DisplayMetrics;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.MotionEvent;
import android.view.SurfaceHolder;
import android.view.SurfaceView;
import android.view.View;
import android.view.ViewGroup;
import android.view.WindowManager;
import android.widget.Button;
import android.widget.EditText;
import android.widget.RelativeLayout;
import android.widget.SeekBar;
import android.widget.TextView;

import com.android.logcat.log.ALog;
import com.bowoon.android.android_videoview.R;
import com.bowoon.android.android_videoview.gif.GIFExtractor;
import com.bowoon.android.android_videoview.gif.GIFView;
import com.bowoon.android.android_videoview.vo.Item;

import java.io.IOException;

import butterknife.ButterKnife;
import butterknife.OnClick;

public class VideoPlayerActivity extends Activity implements SurfaceHolder.Callback {
    private final String TAG = "VideoPlayerActivity";
    private SurfaceView mSurfaceView;
    private SurfaceHolder mSurfaceHolder;
    private MediaPlayer mMediaPlayer;
    private RelativeLayout mVideoInfoView;
    private TextView mVideoTitle, mVideoTime;
    private SeekBar mVideoTimeSeekBar;
    private boolean mSeekBarFlag, isPlay;
    private Item mVideoItem;
    private DisplayMetrics mDisplayMetrics;
    private Button mPlayBtn, mPauseBtn, mDialogBtn, mServiceBtn;
    private long mStartTime, mEndTime;
    private int mFps;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.video_surfaceview);

        setRequestedOrientation(ActivityInfo.SCREEN_ORIENTATION_SENSOR);
        getWindow().addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON);

        Intent intent = getIntent();
        mVideoItem = (Item) intent.getSerializableExtra("videoContent");

        initView();
        registerListener();
    }

    @Override
    public void onConfigurationChanged(Configuration newConfig) {
        super.onConfigurationChanged(newConfig);
        Log.i(TAG, "onConfigurationChanged");

        arrangeVideo();
    }

    private void initView() {
        ButterKnife.bind(this);
        mMediaPlayer = new MediaPlayer();
        mVideoInfoView = (RelativeLayout) findViewById(R.id.video_information);
        mVideoTitle = (TextView) findViewById(R.id.play_video_title);
        mVideoTime = (TextView) findViewById(R.id.video_time);
        mSurfaceView = (SurfaceView) findViewById(R.id.main_surfaceview);
        mVideoTimeSeekBar = (SeekBar) findViewById(R.id.video_seekbar);
        mPlayBtn = (Button) findViewById(R.id.video_play);
        mPauseBtn = (Button) findViewById(R.id.video_pause);
        mDialogBtn = (Button) findViewById(R.id.make_gif);
        mServiceBtn = (Button) findViewById(R.id.video_service);
        mSurfaceHolder = mSurfaceView.getHolder();
        mSurfaceHolder.addCallback(this);
        mDisplayMetrics = getApplicationContext().getResources().getDisplayMetrics();

        mVideoTitle.setText(mVideoItem.getTitle());
    }

    private void registerListener() {
        mVideoTimeSeekBar.setOnSeekBarChangeListener(new SeekBar.OnSeekBarChangeListener() {
            @Override
            public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {
                if (seekBar.getMax() == progress) {
                    mMediaPlayer.stop();
                    finish();
                }
            }

            @Override
            public void onStartTrackingTouch(SeekBar seekBar) {
                mSeekBarFlag = false;
            }

            @Override
            public void onStopTrackingTouch(SeekBar seekBar) {
                mSeekBarFlag = true;
                int position = seekBar.getProgress();
                mMediaPlayer.seekTo(position);
                new ProgressSeekBar().start();
            }
        });
    }

    @Override
    public boolean onTouchEvent(MotionEvent event) {
        if (event.getAction() == MotionEvent.ACTION_UP) {
            mVideoInfoView.setVisibility(View.VISIBLE);
            mVideoTitle.setVisibility(View.VISIBLE);
//            mDialogBtn.setVisibility(View.VISIBLE);
            mServiceBtn.setVisibility(View.VISIBLE);
            mVideoTime.setText(getStringTime(mMediaPlayer.getCurrentPosition()) + " / " + getStringTime(mMediaPlayer.getDuration()));

            new Handler().postDelayed(new Runnable() {
                @Override
                public void run() {
                    mVideoInfoView.setVisibility(View.GONE);
                    mVideoTitle.setVisibility(View.GONE);
//                    mDialogBtn.setVisibility(View.GONE);
                    mServiceBtn.setVisibility(View.GONE);
                }
            }, 5000);

            return false;
        }
        return super.onTouchEvent(event);
    }

    @OnClick({R.id.make_gif, R.id.video_play, R.id.video_pause, R.id.video_service})
    public void buttonOnClickMethod(View view) {
        switch (view.getId()) {
            case R.id.video_service:
                ALog.i("startService");
                Intent serviceIntent = new Intent(getApplicationContext(), VideoService.class);
                serviceIntent.putExtra("video", mVideoItem);
                serviceIntent.putExtra("currentTime", mMediaPlayer.getCurrentPosition());
                ALog.i(mMediaPlayer.getCurrentPosition());
                startService(serviceIntent);
                finish();
                break;
            case R.id.make_gif:
                mMediaPlayer.pause();
                mSeekBarFlag = false;
                AlertDialog.Builder builder = new AlertDialog.Builder(VideoPlayerActivity.this);
                LayoutInflater inflater = getLayoutInflater();
                View v = inflater.inflate(R.layout.dialog_layout, null);
                builder.setView(v);
                Button make = (Button) v.findViewById(R.id.gif_extractor);
                final EditText start = (EditText) v.findViewById(R.id.gif_start);
                final EditText end = (EditText) v.findViewById(R.id.gif_end);
                final EditText fps = (EditText) v.findViewById(R.id.gif_fps);
                final AlertDialog dialog = builder.create();
                make.setOnClickListener(new View.OnClickListener() {
                    @Override
                    public void onClick(View v) {
                        mStartTime = Long.valueOf(start.getText().toString());
                        mEndTime = Long.valueOf(end.getText().toString());
                        mFps = Integer.valueOf(fps.getText().toString());

                        GIFExtractor extractor = new GIFExtractor(getApplicationContext());

                        Intent intent = new Intent(VideoPlayerActivity.this, GIFView.class);
                        intent.putExtra("gifImage", extractor.makeGIF(mVideoItem.getPath(), mStartTime, mEndTime, mFps));
                        startActivity(intent);

//                            mMediaPlayer.start();
//                            mSeekBarFlag = true;
//                            new ProgressSeekBar().start();
                        dialog.dismiss();
                    }
                });

                ViewGroup.LayoutParams params = dialog.getWindow().getAttributes();
                params.width = ViewGroup.LayoutParams.MATCH_PARENT;
                params.height = ViewGroup.LayoutParams.MATCH_PARENT;
                dialog.getWindow().setAttributes((android.view.WindowManager.LayoutParams) params);

                dialog.show();
                break;
            case R.id.video_play:
                if (!isPlay) {
                    isPlay = true;
                    mMediaPlayer.start();
                }
                break;
            case R.id.video_pause:
                if (isPlay) {
                    isPlay = false;
                    mMediaPlayer.pause();
                }
                break;
            default:
                break;
        }
    }

    // 영상 비율에 맞게 확대 / 축소
    private void arrangeVideo() {
        int videoWidth = mMediaPlayer.getVideoWidth();
        int videoHeight = mMediaPlayer.getVideoHeight();

        int screenWidth = mDisplayMetrics.widthPixels;
        int screenHeight = mDisplayMetrics.heightPixels;

        ViewGroup.LayoutParams lp = mSurfaceView.getLayoutParams();

        if (screenWidth < screenHeight) {
            lp.width = screenWidth;
            lp.height = (int) (((float) videoHeight / (float) videoWidth) * (float) screenWidth);
        } else {
            lp.width = (int) (((float) videoWidth / (float) videoHeight) * (float) screenHeight);
            lp.height = screenHeight;
        }

        mSurfaceView.setLayoutParams(lp);
    }

    private void playVideo(String path) {
        try {
            if (mMediaPlayer == null) {
                mMediaPlayer = new MediaPlayer();
            }
            mMediaPlayer.setDataSource(path);
            mMediaPlayer.setDisplay(mSurfaceHolder);
            mMediaPlayer.prepare();
            mMediaPlayer.start();
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    private void releaseMediaPlayer() {
        getWindow().clearFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON);
        mSeekBarFlag = false;
        if (mMediaPlayer != null) {
            mMediaPlayer.release();
            mMediaPlayer = null;
        }
    }

    @Override
    protected void onPause() {
        super.onPause();
        Log.i(TAG, "onPause");
    }

    @Override
    protected void onResume() {
        getWindow().addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON);
        super.onResume();
        Log.i(TAG, "onResume");
    }

    @Override
    protected void onStop() {
        releaseMediaPlayer();
        super.onStop();
    }

    @Override
    protected void onDestroy() {
        releaseMediaPlayer();
        super.onDestroy();
    }

    @Override
    public void surfaceCreated(SurfaceHolder holder) {
        Log.i(TAG, "surfaceCreated");
        mSeekBarFlag = true;
        isPlay = true;
        playVideo(mVideoItem.getPath());
        mVideoTimeSeekBar.setMax(mMediaPlayer.getDuration());
        new ProgressSeekBar().start();
    }

    @Override
    public void surfaceChanged(SurfaceHolder holder, int format, int width, int height) {
        Log.i(TAG, "surfaceChanged");
        if (mMediaPlayer != null) {
            mMediaPlayer.setDisplay(holder);
            arrangeVideo();
        }
    }

    @Override
    public void surfaceDestroyed(SurfaceHolder holder) {
        Log.i(TAG, "surfaceDestroyed");
    }

    private String getStringTime(int time) {
        int currentSecond = time / 1000;
        int second = currentSecond % 60;
        int minute = (currentSecond / 60) % 60;
        int hour = currentSecond / 3600;

        return hour + ":" + minute + ":" + second;
    }

    private class ProgressSeekBar extends Thread {
        @Override
        public void run() {
            while (mSeekBarFlag) {
                mVideoTimeSeekBar.setProgress(mMediaPlayer.getCurrentPosition());
            }
        }
    }
}

File: AndroidVideoView/app/src/main/java/com/bowoon/android/android_videoview/video/VideoPlayerActivity.kt
package com.bowoon.android.android_videoview.video

import android.app.Activity
import android.content.Intent
import android.content.pm.ActivityInfo
import android.content.res.Configuration
import android.media.MediaPlayer
import android.os.Bundle
import android.os.Handler
import android.view.MotionEvent
import android.view.SurfaceHolder
import android.view.View
import android.view.WindowManager
import android.widget.SeekBar
import android.widget.Toast
import com.android.logcat.log.ALog
import com.bowoon.android.android_videoview.R
import com.bowoon.android.android_videoview.model.Video
import kotlinx.android.synthetic.main.video_surfaceview.*
import java.io.IOException

class VideoPlayerActivity : Activity() {
    private var mMediaPlayer: MediaPlayer = MediaPlayer()
    private var mSeekBarFlag = false
    private var isPlay = false
    private lateinit var mVideoVideo: Video
    
    companion object {
        const val LANDSCAPE = 1
        const val PORTRAIT = 0
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.video_surfaceview)

        ALog.i("VideoPlayerActivity")

        requestedOrientation = ActivityInfo.SCREEN_ORIENTATION_SENSOR
        window.addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON)

        mVideoVideo = intent.getSerializableExtra("videoContent") as Video

        initView()
        registerListener()
    }

    override fun onConfigurationChanged(newConfig: Configuration) {
        super.onConfigurationChanged(newConfig)
        ALog.i("onConfigurationChanged")

        resizeSurfaceView()
    }

    private fun resizeSurfaceView() {
        val mDisplayMetrics = applicationContext.resources.displayMetrics

        if (windowManager.defaultDisplay.rotation == LANDSCAPE) {
            Toast.makeText(this, "landscape", Toast.LENGTH_SHORT).show()
            mainSurfaceView.layoutParams.apply {
                width = mDisplayMetrics.widthPixels
                height = mDisplayMetrics.heightPixels
            }
        } else if (windowManager.defaultDisplay.rotation == PORTRAIT) {
            Toast.makeText(this, "portrait", Toast.LENGTH_SHORT).show()
            mainSurfaceView.layoutParams.apply {
                width = mDisplayMetrics.widthPixels
                height = (mMediaPlayer.videoHeight.toFloat() / mMediaPlayer.videoWidth.toFloat() * mDisplayMetrics.widthPixels.toFloat()).toInt()
            }
        }
    }

    private fun initView() {
        mainSurfaceView.holder.addCallback(SurfaceViewCallbackClass())
        playVideoTitle.text = mVideoVideo.title
    }

    private fun registerListener() {
        videoSeekbar.setOnSeekBarChangeListener(object : SeekBar.OnSeekBarChangeListener {
            override fun onProgressChanged(seekBar: SeekBar, progress: Int, fromUser: Boolean) {
                if (seekBar.max == progress) {
                    mMediaPlayer.stop()
                    finish()
                }
            }

            override fun onStartTrackingTouch(seekBar: SeekBar) {
                mSeekBarFlag = false
            }

            override fun onStopTrackingTouch(seekBar: SeekBar) {
                mSeekBarFlag = true
                val position = seekBar.progress
                mMediaPlayer.seekTo(position)
                ProgressSeekBar().start()
            }
        })

        videoService.setOnClickListener {
            ALog.i("startService")
            startService(Intent(applicationContext, VideoService::class.java).apply {
                putExtra("video", mVideoVideo)
                putExtra("currentTime", mMediaPlayer.currentPosition)
            })
            finish()
        }

        videoPlay.setOnClickListener {
            if (!isPlay) {
                isPlay = true
                mMediaPlayer.start()
            }
        }

        videoPause.setOnClickListener {
            if (isPlay) {
                isPlay = false
                mMediaPlayer.pause()
            }
        }
    }

    override fun onTouchEvent(event: MotionEvent): Boolean {
        if (event.action == MotionEvent.ACTION_UP) {
            videoInformation.visibility = View.VISIBLE
            playVideoTitle.visibility = View.VISIBLE
//            mDialogBtn.visibility = View.VISIBLE
            videoService.visibility = View.VISIBLE
            videoTime.text = String.format("%s / %s", getStringTime(mMediaPlayer.currentPosition), getStringTime(mMediaPlayer.duration))

            Handler().postDelayed({
                videoInformation.visibility = View.GONE
                playVideoTitle.visibility = View.GONE
//                mDialogBtn.visibility = View.GONE
                videoService.visibility = View.GONE
            }, 5000)

            return false
        }
        return super.onTouchEvent(event)
    }

    private fun playVideo(path: String) {
        try {
            mMediaPlayer.apply {
                setDataSource(path)
                setDisplay(mainSurfaceView.holder)
                prepare()
                start()
            }
        } catch (e: IOException) {
            e.printStackTrace()
        }

    }

    private fun releaseMediaPlayer() {
        window.clearFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON)
        mSeekBarFlag = false
        mMediaPlayer.release()
    }

    override fun onPause() {
        super.onPause()
        ALog.i("onPause")
    }

    override fun onResume() {
        window.addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON)
        super.onResume()
        ALog.i("onResume")
    }

    override fun onStop() {
        releaseMediaPlayer()
        super.onStop()
    }

    override fun onDestroy() {
        releaseMediaPlayer()
        super.onDestroy()
    }

    private fun getStringTime(time: Int): String {
        val currentSecond = time / 1000
        val second = currentSecond % 60
        val minute = currentSecond / 60 % 60
        val hour = currentSecond / 3600

        return "$hour:$minute:$second"
    }

    private inner class ProgressSeekBar : Thread() {
        override fun run() {
            while (mSeekBarFlag) {
                videoSeekbar.progress = mMediaPlayer.currentPosition
            }
        }
    }

    private inner class SurfaceViewCallbackClass : SurfaceHolder.Callback {
        override fun surfaceCreated(holder: SurfaceHolder) {
            ALog.i("surfaceCreated")
            mSeekBarFlag = true
            isPlay = true
            playVideo(mVideoVideo.path)
            videoSeekbar.max = mMediaPlayer.duration
            ProgressSeekBar().start()
        }

        override fun surfaceChanged(holder: SurfaceHolder, format: Int, width: Int, height: Int) {
            ALog.i("surfaceChanged")
            mMediaPlayer.setDisplay(holder)
            resizeSurfaceView()
        }

        override fun surfaceDestroyed(holder: SurfaceHolder) {
            ALog.i("surfaceDestroyed")
        }
    }
}
